File: 02. User/Sdk.User/UserCli.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Api.Auth;
using Api.User;
using static Api.User.IUserApi;

namespace Sdk.User
{
    public class UserCli : IUserApi
    {
        /// <summary>
        /// Creates a client tool for accessing the web service
        /// located in the specified URL.
        /// </summary>
        /// <param name="serviceBaseUrl">Base URL where the web service is serving</param>
        /// <param name="authCli">Auth service client library. Required to include authentication header</param>
        public UserCli(string serviceBaseUrl, IAuthApi authCli)
        {
            _base = serviceBaseUrl;
            _client = new HttpClient();
            _authCli = authCli;
        }

        /// <summary>
        /// Creates a client tool for accessing the web service
        /// for automatic tests purposes. The HttpClient object is passed
        /// so that the integration test cases can make use of the testing
        /// client tool instead of the actual one.
        /// </summary>
        /// <param name="testClient">Alternative HttpClient tool to be used</param>
        /// <param name="authCli">Auth service client library. Required to include authentication header</param>
        public UserCli(HttpClient testClient, IAuthApi authCli)
        {
            _base = "";
            _client = testClient;
            _authCli = authCli;
        }

        public async Task<List<UserInfo>> GetUsers(string groupId)
        {
            // Send request.
            ArrangeAuthenticatedRequest();
            HttpResponseMessage response = await _client.GetAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{USERS_PATH}/"+
                    $"{groupId}/");

            // Provide success.
            if (response.IsSuccessStatusCode)
            {
                if (response.StatusCode == HttpStatusCode.NoContent) return null;
                return await response.Content.ReadAsAsync<List<UserInfo>>();
            }

            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

        public async Task<UserInfo> GetUser(string userId)
        {
            // Send request.
            ArrangeAuthenticatedRequest();
            HttpResponseMessage response = await _client.GetAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{USER_PATH}/" +
                    $"{userId}");

            // Provide success.
            if (response.IsSuccessStatusCode)
            {
                if (response.StatusCode == HttpStatusCode.NoContent) return null;
                return await response.Content.ReadAsAsync<UserInfo>();
            }
            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

        public async Task<GroupInfo> GetGroup(string groupId)
        {
            // Send request.
            ArrangeAuthenticatedRequest();
            HttpResponseMessage response = await _client.GetAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{GROUP_PATH}/" +
                    $"{GROUP_ID_PATH}");

            // Provide success.
            if (response.IsSuccessStatusCode)
            {
                if (response.StatusCode == HttpStatusCode.NoContent) return null;
                return await response.Content.ReadAsAsync<GroupInfo>();
            }
            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

        public async Task<GroupInfo> GetGroupByName(string groupId)
        {
            // Send request.
            ArrangeAuthenticatedRequest();
            HttpResponseMessage response = await _client.GetAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{GROUP_PATH}/" +
                    $"{GROUP_NAME}");

            // Provide success.
            if (response.IsSuccessStatusCode)
            {
                if (response.StatusCode == HttpStatusCode.NoContent) return null;
                return await response.Content.ReadAsAsync<GroupInfo>();
            }
            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

        public async Task<GroupInfo> PostGroup(GroupInfo group)
        {
           // Send request.
            ArrangeAuthenticatedRequest();
            StringContent data = new StringContent(JsonSerializer.Serialize<GroupInfo>(group), Encoding.UTF8, "application/json");
            HttpResponseMessage response = await _client.PostAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{GROUP_PATH}", data);

            // Provide success.
            if (response.IsSuccessStatusCode)
            {
                if (response.StatusCode == HttpStatusCode.NoContent) return null;
                return await response.Content.ReadAsAsync<GroupInfo>();
            }
            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

        public async Task AddToGroup(string userId, GroupInfo group)
        {
           // Send request.
            ArrangeAuthenticatedRequest();
            StringContent data = new StringContent(JsonSerializer.Serialize<GroupInfo>(group), Encoding.UTF8, "application/json");
            HttpResponseMessage response = await _client.PutAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{GROUP_ADD}/" +
                    $"{USER_ID_PATH}", data);

            // Provide success.
            if (response.IsSuccessStatusCode){ return; }
            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

        public async Task<RoleInfo> PostRole(RoleInfo role)
        {
           // Send request.
            ArrangeAuthenticatedRequest();
            StringContent data = new StringContent(JsonSerializer.Serialize<RoleInfo>(role), Encoding.UTF8, "application/json");
            HttpResponseMessage response = await _client.PostAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{ROLE_PATH}", data);

            // Provide success.
            if (response.IsSuccessStatusCode)
            {
                if (response.StatusCode == HttpStatusCode.NoContent) return null;
                return await response.Content.ReadAsAsync<RoleInfo>();
            }
            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

         public async Task AddToRole(string userId, RoleInfo role)
        {
           // Send request.
            ArrangeAuthenticatedRequest();
            StringContent data = new StringContent(JsonSerializer.Serialize<RoleInfo>(role), Encoding.UTF8, "application/json");
            HttpResponseMessage response = await _client.PutAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{ROLE_ADD}/" +
                    $"{USER_ID_PATH}", data);

            // Provide success.
            if (response.IsSuccessStatusCode){ return; }
            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

        public async Task<UserInfo> PostUser(UserInfo user)
        {
           // Send request.
            ArrangeAuthenticatedRequest();
            StringContent data = new StringContent(JsonSerializer.Serialize<UserInfo>(user), Encoding.UTF8, "application/json");
            HttpResponseMessage response = await _client.PostAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{USER_PATH}", data);

            // Provide success.
            if (response.IsSuccessStatusCode)
            {
                if (response.StatusCode == HttpStatusCode.NoContent) return null;
                return await response.Content.ReadAsAsync<UserInfo>();
            }
            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

        public async Task<UserInfo> PutUser(UserInfo user, string userId)
        {
            ArrangeAuthenticatedRequest();
            StringContent data = new StringContent(JsonSerializer.Serialize<UserInfo>(user), Encoding.UTF8, "application/json");
            HttpResponseMessage response = await _client.PutAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{USER_PATH}/" +
                    $"{userId}", data);

            // Provide success.
            if (response.IsSuccessStatusCode)
            {
                if (response.StatusCode == HttpStatusCode.NoContent) return null;
                return await response.Content.ReadAsAsync<UserInfo>();
            }
            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

        public async Task<bool> DeleteUser(string userId)
        {
            ArrangeAuthenticatedRequest();
            HttpResponseMessage response = await _client.DeleteAsync($"{_base}/" +
                    $"{IUserApi.SERVICE_ROUTE}/" +
                    $"{USER_PATH}/" +
                    $"{userId}");

            // Provide success.
            if (response.IsSuccessStatusCode){ return true; }
            // Error.
            throw new Exception(await response.Content.ReadAsStringAsync());
        }

        /// <summary>
        /// Overrides the session token as provided by the auth client by using the passed one.
        /// This is used by calls from service to service transferring the session from call to call.
        /// </summary>
        public void UseSession(string jwtToken)
        {
            _jwtTokenOverridden = jwtToken;
        }

        /// <summary>
        /// Places in the request header for the next remote calls the authentication
        /// tokens with the help of the Auth SDK that holds the current session tokens,
        /// or using the overridden one if present.
        /// </summary>
        private void ArrangeAuthenticatedRequest()
        {
            // Overridden session JWT token.
            if (_jwtTokenOverridden != null) _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _jwtTokenOverridden);

            // Using the session as provided by the inyected auth service client.
            if (_authCli == null) return;
            _authCli.SetSession(_client);
        }
    
        private HttpClient _client;
        private string _base;
        private IAuthApi _authCli;
        private string _jwtTokenOverridden;
    }
}

File: ContentManagement-Frontend/ClientApp/src/sdk/UserCli.js
import AuthCli from "./AuthCli"
/**
 * JavaScript version of the User service API
 */
export default class UserCli {

  //
  // Path constants
  //
  static get SERVICE_ROUTE() { return "user" }
  static get USERS_PATH() { return "users" }
  static get GROUP_ID_PATH() { return "groupId" }
  static get USER_PATH() { return "user" }
  static get USER_ID_PATH() { return "userId" }
  static get GROUPS_PATH() { return "groups" }
  static get GROUP_PATH() { return "group" }
  static get GROUP_NAME() { return "groupName" }
  static get GROUP_ADD() { return "addToGroup" }
  static get ROLES_PATH() { return "roles" }
  static get ROLE_PATH() { return "role" }
  static get ROLE_ADD() { return "addToRole" }
  static get ROLE_ID_PATH() { return "roleId" }
  static get ROLE_NAME() { return "groupName" }

  static get prefix() {return ('http://localhost:8002/')}
  /**
   * Provides detailed information about all the users. In case of groupId is not null, it will be provide the user that belong to the group.
   * @param {string} groupId Unique internal identifier of the group
   * @returns 
   */
  static async GetUsers(groupId){
    var verb = "GET"
    var path = this.SERVICE_ROUTE + "/" + this.USERS_PATH + "?" + this.GROUP_ID_PATH + "=" + groupId
    var body = null
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }
  
  /**
   * Provides detailed information about the user requested
   * @param {string} userId Unique internal identifier of the user
   * @returns 
   */
  static async GetUser(userId){
    var verb = "GET"
    var path = this.SERVICE_ROUTE + "/" + this.USER_PATH + "?" + this.USER_ID_PATH + "=" + userId
    var body = null
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }

  /**
   * Provides detailed information about the groups in the system
   * @returns 
   */
  static async GetGroups(){
    var verb = "GET"
    var path = this.SERVICE_ROUTE + "/" + this.GROUPS_PATH
    var body = null
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }

  /**
   * Provides detailed information about the user group requested
   * @param {string} groupId Unique internal identifier of the group
   * @returns 
   */
  static async GetGroup(groupId){
    var verb = "GET"
    var path = this.SERVICE_ROUTE + "/" + this.GROUP_PATH + "?" + this.GROUP_ID_PATH + "=" + groupId
    var body = null
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }

  /**
   * Provides detailed information about the user group requested
   * @param {string} groupName Unique internal identifier name of the group
   * @returns 
   */
  static async GetGroupByName(groupName){
    var verb = "GET"
    var path = this.SERVICE_ROUTE + "/" + this.GROUP_PATH + "?" + this.GROUP_NAME + "=" + groupName
    var body = null
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }

  /**
   * Save detailed information about group
   * @param {GroupInfo} group Group entity that contains the information
   * @returns 
   */
  static async PostGroup(group){
    var verb = "POST"
    var path = this.SERVICE_ROUTE + "/" + this.GROUP_PATH
    var body = group
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }

  /**
   * Assign user to a given group
   * @param {string} userId Unique internal identifier of the user
   * @param {GroupInfo} group Group entity that contains the information
   * @returns 
   */
  static async AddToGroup(userId, group){
    var verb = "PUT"
    var path = this.SERVICE_ROUTE + "/" + this.GROUP_ADD + "?" + this.USER_ID_PATH + "=" + userId
    var body = group
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }

    /**
   * Provides detailed information about the roles in the system
   * @returns 
   */
    static async GetRoles(){
      var verb = "GET"
      var path = this.SERVICE_ROUTE + "/" + this.ROLES_PATH
      var body = null
    
      return AuthCli.Fetch(this.prefix + path, verb, body)
    }
  
    /**
     * Provides detailed information about the role requested
     * @param {string} roleId Unique internal identifier of the role
     * @returns 
     */
    static async GetRole(roleId){
      var verb = "GET"
      var path = this.SERVICE_ROUTE + "/" + this.ROLE_PATH + "?" + this.ROLE_ID_PATH + "=" + roleId
      var body = null
    
      return AuthCli.Fetch(this.prefix + path, verb, body)
    }
  
    /**
     * Provides detailed information about the role requested
     * @param {string} groupName Unique internal identifier name of the role
     * @returns 
     */
    static async GetRoleByName(roleName){
      var verb = "GET"
      var path = this.SERVICE_ROUTE + "/" + this.ROLE_PATH + "?" + this.ROLE_NAME + "=" + roleName
      var body = null
    
      return AuthCli.Fetch(this.prefix + path, verb, body)
    }

  /**
   * Save detailed information about role
   * @param {RoleInfo} role Role entity that contains the information
   * @returns 
   */
  static async PostRole(role){
    var verb = "POST"
    var path = this.SERVICE_ROUTE + "/" + this.ROLE_PATH
    var body = role
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }

  /**
   * Assign user to a given role
   * @param {*} userId Unique internal identifier of the user
   * @param {*} role Role entity that contains the information
   * @returns 
   */
  static async AddToRole(userId, role){
    var verb = "PUT"
    var path = this.SERVICE_ROUTE + "/" + this.ROLE_ADD + "?" + this.USER_ID_PATH + "=" + userId
    var body = role
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }

  /**
   * Save detailed information about user
   * @param {UserInfo} user User entity that contains the information
   * @returns 
   */
  static async PostUser(user){
    var verb = "POST"
    var path = this.SERVICE_ROUTE + "/" + this.USER_PATH
    var body = user
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }

  /**
   * Modifies user who already exist in the db
   * @param {UserInfo} user User entity that contains the new information
   * @param {string} userId Unique internal identifier of the user who is going to be replaced
   * @returns 
   */
  static async PutUser(user, userId){
    var verb = "PUT"
    var path = this.SERVICE_ROUTE + "/" + this.USER_PATH + "?" + this.USER_ID_PATH + "=" + userId
    var body = user
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }

  /**
   * Removes from the system the information about user
   * @param {string} userId Unique internal identifier of the user to be deleted
   * @returns 
   */
  static async DeleteUser(userId) {
    var verb = "DELETE"
    var path = this.SERVICE_ROUTE + "/" + this.USER_PATH + "?" + this.USER_ID_PATH + "=" + userId
    var body = null
  
    return AuthCli.Fetch(this.prefix + path, verb, body)
  }
}